File: app/src/main/java/me/yifeiyuan/flapdev/MainActivity.java
package me.yifeiyuan.flapdev;

import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;

import java.util.ArrayList;
import java.util.List;

import me.yifeiyuan.flap.FlapAdapter;
import me.yifeiyuan.flap.ktmodule.KtComponentModel;
import me.yifeiyuan.flapdev.components.customviewtype.CustomModel;
import me.yifeiyuan.flapdev.components.databindingsample.SimpleDataBindingModel;
import me.yifeiyuan.flapdev.components.generictest.GenericModel;
import me.yifeiyuan.flapdev.components.simpleimage.SimpleImageModel;
import me.yifeiyuan.flapdev.components.simpletext.SimpleTextModel;
import me.yifeiyuan.ktx.foundation.othermodule.JavaModuleModel;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView recyclerView = findViewById(R.id.rv_items);

//        createSimpleTestCase(recyclerView);
        createAdvanceTestCase(recyclerView);
    }

    private void createSimpleTestCase(final RecyclerView recyclerView) {
        FlapAdapter adapter = new FlapAdapter();

        List<Object> models = new ArrayList<>();

        models.add(new SimpleTextModel("Android"));
        models.add(new SimpleTextModel("Java"));
        models.add(new SimpleTextModel("Kotlin"));

        models.add(new SimpleImageModel());
        models.add(new SimpleImageModel());
        models.add(new SimpleImageModel());
        models.add(new SimpleImageModel());

        adapter.setData(models);

        recyclerView.setAdapter(adapter);
    }

    private void createAdvanceTestCase(final RecyclerView recyclerView) {

        List<Object> models = mockModels();

        FlapAdapter adapter = new FlapAdapter();
        adapter.setUseComponentPool(true)
                .setLifecycleEnable(true)
                .setLifecycleOwner(this)
                .setData(models);

        recyclerView.setAdapter(adapter);
    }

    private List<Object> mockModels() {

        List<Object> models = new ArrayList<>();
        models.add(new SimpleTextModel("Flap（灵动）"));
        models.add(new SimpleTextModel("一个基于 RecyclerView 的页面组件化框架"));
        models.add(new SimpleTextModel("—— by 程序亦非猿"));

        models.add(new SimpleImageModel());

        models.add(new CustomModel());
        models.add(new GenericModel());
        models.add(new SimpleDataBindingModel());

        models.add(new JavaModuleModel());
        models.add(new KtComponentModel());

        return models;
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.main,menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.differ:
                startActivity(new Intent(MainActivity.this,DifferActivity.class));
                break;

            case R.id.kotlin:
                startActivity(new Intent(MainActivity.this,KotlinTestActivity.class));
                break;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/me/yifeiyuan/flapdev/MainActivity.kt
package me.yifeiyuan.flapdev

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.navigation.NavigationView
import me.yifeiyuan.flap.ext.ifIs
import me.yifeiyuan.flapdev.testcases.*
import java.lang.Exception

class MainActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "MainActivity"
    }

    lateinit var mainViewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main2)

        val toolbar: Toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val drawerLayout: DrawerLayout = findViewById(R.id.drawer_layout)
        val navView: NavigationView = findViewById(R.id.nav_view)
        navView.setItemIconTintList(null)
        val toggle = ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.nav_app_bar_open_drawer_description, R.string.navigation_drawer_close)
        drawerLayout.setDrawerListener(toggle)
        toggle.syncState()

        navView.setNavigationItemSelectedListener { it ->

            var title = "Flap"
            var subtitle = ""
            when (it.itemId) {
                R.id.nav_multi_type -> {
                    title = "多类型"
                    subtitle = "多 itemViewType"
                    replace(MultiTypeTestcase::class.java)
                }
                R.id.nav_preload -> {
                    title = "预加载"
                    subtitle = "在滑动到底部或顶部之前预先加载"
                    replace(PreloadTestcase::class.java)
                }
                R.id.nav_component_pool -> {
                    title = "ComponentPool"
                    subtitle = "滑动列表观察缓存情况"
                    replace(ComponentPoolTestcase::class.java)
                }
                R.id.nav_flap_rv -> {
                    title = "FlapRecyclerView"
                    subtitle = "自定义 RecyclerView"
                    replace(FlapRecyclerViewTestcase::class.java)
                }
                R.id.nav_viewpager2 -> {
                    title = "ViewPager2 适配"
                    subtitle = "ViewPager2+FlapAdapter"
                    replace(ViewPager2Testcase::class.java)
                }
                R.id.nav_diff -> {
                    title = "FlapDifferAdapter"
                    subtitle = "修改数据后，下拉刷新观察"
                    replace(FlapDifferAdapterTestcase::class.java)
                }
                R.id.nav_header_footer -> {
                    subtitle = "Header、Footer"
                    replace(HeaderFooterTestcase::class.java)
                }
                R.id.nav_animation -> {
                    subtitle = "动画"
                    replace(AnimationTestcase::class.java)
                }
                R.id.nav_github_demo -> {
                    subtitle = "GitHub Demo"
                    replace(GitHubDemoFragment::class.java)
                }
                R.id.nav_item_decorations -> {
                    subtitle = "ItemDecorations"
                    replace(ItemDecorationTestcase::class.java)
                }
            }
            drawerLayout.close()

            toolbar.title = title
            toolbar.subtitle = subtitle

            true
        }

        toolbar.title = "多类型"
        toolbar.subtitle = "多 itemViewType"
        replace(MultiTypeTestcase::class.java)

        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener {
            val fragment = supportFragmentManager.findFragmentByTag("testcase")
            fragment.ifIs<IMenuView> {
                showMenu()
            }
        }

        fab.setOnLongClickListener {
            val fragment = supportFragmentManager.findFragmentByTag("testcase")
            if (fragment is Scrollable) {
                fragment.scrollToTop()
            }
            true
        }
    }

    private fun <T : Fragment> replace(fragmentClass: Class<T>, args: Bundle? = null) {
        supportFragmentManager.beginTransaction().replace(R.id.container, fragmentClass, args, "testcase").commitAllowingStateLoss()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.author -> {
                val url = "https://www.yuque.com/cxyfy/blog/axzri0"
                val uri: Uri = Uri.parse(url)
                val intent = Intent(Intent.ACTION_VIEW, uri)
                try {
                    startActivity(intent)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

            R.id.github -> {
                val url = "https://github.com/AlanCheen/Flap"
                val uri: Uri = Uri.parse(url)
                val intent = Intent(Intent.ACTION_VIEW, uri)
                try {
                    startActivity(intent)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.home, menu)
        return true
    }
}